collectors/modality-probe-debug-collector/src/cli.rs

```diff
@@ -0,0 +1,550 @@
+use err_derive::Error;
+use std::convert::TryFrom;
+use std::fs::File;
+use std::io::prelude::*;
+use std::net::SocketAddrV4;
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+use goblin::elf::Elf;
+
+use modality_probe_debug_collector::{Config, ProbeAddr, Target, Word};
+
+#[derive(Debug, Error)]
+pub enum CLIError {
+    #[error(display = "Address is not valid or too large \"{}\"", _0)]
+    AddressNotValid(String),
+    #[error(display = "Error opening ELF file")]
+    ElfFileError,
+    #[error(display = "Symbols were given but no elf file was specified")]
+    MissingElfFileError,
+    #[error(display = "Given interval not a valid duration: {}", _0)]
+    InvalidInterval(String),
+    #[error(display = "Symbol not found in given ELF file: \"{}\"", _0)]
+    SymbolNotFound(String),
+}
+
+#[derive(Debug, Default, StructOpt)]
+#[structopt(
+    name = "modality-probe-debug-collector",
+    about = "Periodically collects logs from microcontrollers over debug interfaces; outputs them to a file."
+)]
+pub struct CLIOptions {
+    /// Session id to associate with the collected trace data
+    #[structopt(short = "s", long = "session-id", default_value = "0")]
+    session_id: u32,
+
+    /// Specifies 32 bit architecture of target system
+    #[structopt(short = "32", long = "32-bit", conflicts_with = "word-size-64")]
+    word_size_32: bool,
+
+    /// Specifies 32 bit architecture of target system
+    #[structopt(short = "64", long = "64-bit")]
+    word_size_64: bool,
+
+    /// Path of ELF file for endianness and symbol resolution
+    #[structopt(short = "e", long = "elf", parse(from_os_str))]
+    elf_path: Option<PathBuf>,
+
+    /// Chip type of target system to attach to
+    #[structopt(
+        short = "a",
+        long = "attach",
+        conflicts_with = "gdb-addr",
+        required_unless = "gdb-addr"
+    )]
+    attach: Option<String>,
+
+    /// Address of gdb server to connect to
+    #[structopt(short = "g", long = "gdb-addr", required_unless = "attach")]
+    gdb_addr: Option<SocketAddrV4>,
+
+    /// Interval between log collections. Ex: "2 min 15 sec 500 milli 250 micro"
+    #[structopt(short = "i", long = "interval")]
+    interval: String,
+
+    /// Output file location
+    #[structopt(short = "o", long = "output-file", parse(from_os_str))]
+    output_path: PathBuf,
+
+    /// Symbols and/or raw addresses of probes or probe pointers.
+    /// Raw addresses should be in hex format, prefixed with '0x' or '0X'
+    /// Probe pointer addresses and symbols should be prefixed with `*`.
+    /// Ex: *0X100 0x104 symbol1 *symbol2 0X108 *0x10c
+    probe_syms: Vec<String>,
+}
+
+/// Turn CLI options into configuration for the collector
+pub(crate) fn config_from_options(options: CLIOptions) -> Result<Config, CLIError> {
+    let mut probe_addrs = Vec::new();
+    let mut symbols = Vec::new();
```

Comment:
Minor style nit - these two variables should probably be declared immediately prior to their point of first use.  Shortening the distance  between declaration, initialization, and use reduces the surface area for accidental misuse and reduces no-value scrollback for context discovery.